### ISSUE TITLE
Locals are within scope of nested fn calls!

### ISSUE THREAD

**Issue by dkosbie** [brython-dev/brython#780]
Hi all.  This is a pretty big issue for us...  

In the code below, x should not be visible in g(), but it is.  Plus, as shown further below:  x is visible in g(), but it is not local, not global, and not nonlocal.

Here's some code:
```
def g():
    print(x) # should crash, of course

def f():
    x = 42
    g()

f()

# Expected: NameError: name 'x' is not defined
# Observed: prints 42
```

Here again, but we see in g() that x is not local nor global:
```
def g():
    print('x' in globals()) # Correctly prints False
    print('x' in locals())  # Correctly prints False
    print(x) # But this still prints 42

def f():
    x = 42
    g()

f()
```

Once more, and we see x is also not nonlocal in g():
```
def g():
    nonlocal x # Correctly raises error: no binding for nonlocal 'x' found
    print(x)

def f():
    x = 42
    g()

f()
```

So:  x is visible in g(), but it is not local, not global, and not nonlocal.

We hope you can help with this.  Thanks!!!

David

**Comment by PierreQuentel** [comment]
The commit referenced above actually fixes the bug, but since it's in a branch that is not merged yet with master I keep the issue open. The merge should be done at the end of next week.

**Comment by dkosbie** [comment]
Excellent!  Thanks again, as always.  :-)

David

**Comment by dkosbie** [comment]
Hi Pierre.  Are we still on track for this to be merged soon?  As it turns out, we have to freeze our production pipeline until this particular merge is released...

Thanks!

David

**Comment by PierreQuentel** [comment]
Hi David,

I merge the branch that fixes this bug this morning. It should be fixed now, but it showed me that name resolution has other issues, so I opened another branch, name_resolution, that I should also merge soon. But in the mean time I think you can use the current master.

**Comment by dkosbie** [comment]
Hi Pierre.  This is great, thanks!  We look forward to those future patches, too, but this should get us back in business.  Thanks again!!!

David

**Comment by PierreQuentel** [comment]
@dkosbie 
I have merged branch name_resolution into master. Can you check that the issue is solved and if so, close it ?
Thank you for this report and all the others ; Brython would not be what it is now without all the testing and the feedback of users like you :-)

**Comment by dkosbie** [comment]
Hi Pierre.  Thanks for checking.  Unfortunately, when we use this version
of Brython, we are having broad and significant stability issues (too broad
to give you specific issues, sorry).  We are not sure why.  But for now we
had to rollback to the previous Brython version.  When we have a few spare
cycles, we'll dig further into this.  It may be a few weeks, though.

Thanks.

David

**Comment by PierreQuentel** [comment]
Hi David,

As you certainly saw, I have merged the branch and published a release candidate for the next version 3.5.0 because I thought it might break existing code.

Is there something I can do to find why yours doesn't work anymore ? I suppose it's not open sourced but if there is a way I can get access to it for debugging, please tell me.

**Comment by schmave** [comment]
Hi Pierre,

I've started working with David and I looked into these issues a bit today. It looks to me that the name resolution errors are all fixed. I made PRs for some unrelated regressions that were causing errors in our product. I also made a PR to add back some tests that you removed because it seems like it might have been unintentional to remove them. Let me know what you think.

Evan

**Comment by schmave** [comment]
One more regression that popped up between Feb 7 and now was that the running time of our application increased by about 50%. I think I have fixed that in #831, so this issue can now be closed.
